client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,344 @@
+    _telemetricsThread.join();
+  }
+}
+
+Result TelemetricsHandler::checkHttpResponse(
+    std::unique_ptr<httpclient::SimpleHttpResult> const& response) const {
+  using basics::StringUtils::concatT;
+  using basics::StringUtils::itoa;
+  if (response == nullptr || !response->isComplete()) {
+    return {TRI_ERROR_INTERNAL, this->_httpClient->getErrorMessage()};
+  }
+  if (response->wasHttpError()) {
+    auto errorNum = TRI_ERROR_INTERNAL;
+    std::string errorMsg = response->getHttpReturnMessage();
+    // Handle case of no body:
+    try {
+      auto bodyBuilder = response->getBodyVelocyPack();
+      velocypack::Slice error = bodyBuilder->slice();
+      if (!error.isNone() && error.hasKey(StaticStrings::ErrorMessage)) {
+        errorNum = ErrorCode{
+            error.get(StaticStrings::ErrorNum).getNumericValue<int>()};
+        errorMsg = error.get(StaticStrings::ErrorMessage).copyString();
+      }
+    } catch (...) {
+    }
+    return {errorNum,
+            concatT("got invalid response from server: HTTP ",
+                    itoa(response->getHttpReturnCode()), ": ", errorMsg)};
+  }
+  return {TRI_ERROR_NO_ERROR};
+}
+
+void TelemetricsHandler::fetchTelemetricsFromServer() {
+  _telemetricsFetchedInfo.clear();
+  std::string const url = "/_admin/telemetrics";
+  Result result = {TRI_ERROR_NO_ERROR};
+  uint32_t timeoutInSecs = 1;
+  while (!_server.isStopping()) {
+    ClientManager clientManager(
+        _server.getFeature<HttpEndpointProvider, ClientFeature>(),
+        Logger::FIXME);
+    std::unique_lock lk(_mtx);
+
+    _httpClient = clientManager.getConnectedClient(true, false, false, 0);
+    if (_httpClient != nullptr && _httpClient->isConnected()) {
+      auto& clientParams = _httpClient->params();
+      clientParams.setRequestTimeout(30);
+      lk.unlock();
+      std::unique_ptr<httpclient::SimpleHttpResult> response(
+          _httpClient->request(rest::RequestType::GET, url, nullptr, 0));
+      lk.lock();
+
+      result = this->checkHttpResponse(response);
+      _telemetricsFetchResponse = result;
+      if (auto errorNum = result.errorNumber();
+          errorNum == TRI_ERROR_NO_ERROR ||
+          errorNum == TRI_ERROR_HTTP_FORBIDDEN) {
+        _telemetricsFetchedInfo.add(response->getBodyVelocyPack()->slice());
+        break;
+      }
+    }
+    _runCondition.wait_for(lk, std::chrono::seconds(timeoutInSecs),
+                           [this] { return _server.isStopping(); });
+    timeoutInSecs *= 2;
+    timeoutInSecs = std::min<uint32_t>(100, timeoutInSecs);
+  }
+}
+
+// Sends telemetrics to the endpoint and tests the redirection when telemetrics
+// is sent to an endpoint, though not exactly the same thing, because here the
+// endpoint is local, hence the same client is used and a new endpoint is not
+// created.
+std::optional<VPackBuilder> TelemetricsHandler::sendTelemetricsToEndpoint(
+    std::string const& reqUrl) {
+  VPackBuilder testResponseBuilder;
```

Comment:
why is this variable called `testResponseBuilder`?

Reply:
Because it's the builder from the response that is gonna be sent when the function `sendTelemetricsToEndpoint()` is used to test redirection

Reply:
Yes, but is `test` any relevant in a function that is called `sendTelemetricsToEndpoint`? That function could be called from anywhere, at least in theory. Why limit it artifically and assume the reader of the code knows it is called only from tests? That makes things just puzzling for the reader.
Why not simply name it `responseBuilder`, because that is exactly what this variable is used for. Regardless of whether the function is called from a test or from somewhere else.